### PR TITLE
Implemented Combined-texture for WGSL

### PR DESCRIPTION
**Pull request by jkwak-work**
The implementation in hlsl.meta.slang for combined texture was already committed and it appears to be working fine.
The trouble was more on the layout rule where the synthesized textures and samplers were not getting the binding offset values correctly.

When the combined-texture is used, it is replaced with a texture and a sampler.
For WGSL, both cases get layout info as `LayoutResourceKind::DescriptorTableSlot`.
This is little different from how it is handled in HLSL where Texture got `LayoutResourceKind::ShaderResource` and Sampler got `LayoutResourceKind::SamplerState`

This commit also adds a new layout-rule for WGSL.
It is mostly identical to the layout-rule for GLSL.
The difference is that Combined-texture is handled little differently.
This is because for GLSL, there is an actual combined sampler from GLSL language level and Slang don't need to synthesize one. For WGSL, there isn't combined sampler and Slang has to synthesize one.

### CODE REVIEW

File: source/slang/slang-type-layout.cpp [hunk 1]
```diff
@@ -1816,6 +1856,103 @@ LayoutRulesImpl* MetalLayoutRulesFamilyImpl::getHitAttributesParameterRules()
...
+}
+
+LayoutRulesImpl* WGSLLayoutRulesFamilyImpl::getShaderRecordConstantBufferRules()
+{
+    return &kGLSLShaderRecordLayoutRulesImpl_;
+}
+
+LayoutRulesImpl* WGSLLayoutRulesFamilyImpl::getTextureBufferRules(CompilerOptionSet& compilerOptions)
+{
+    if (compilerOptions.shouldUseScalarLayout())
```
Comment (csyonghe) [inline source/slang/slang-type-layout.cpp:1898]: I don't think it is allowed in wgsl to declare a layout that isn't consistent with std140/std430.

File: source/slang/slang-type-layout.cpp [hunk 2]
```diff
@@ -1816,6 +1856,103 @@ LayoutRulesImpl* MetalLayoutRulesFamilyImpl::getHitAttributesParameterRules()
...
+    return &kGLSLVaryingOutputLayoutRulesImpl_;
+}
+
+LayoutRulesImpl* WGSLLayoutRulesFamilyImpl::getSpecializationConstantRules()
+{
+    return &kGLSLSpecializationConstantLayoutRulesImpl_;
+}
+
+LayoutRulesImpl* WGSLLayoutRulesFamilyImpl::getShaderStorageBufferRules(CompilerOptionSet& compilerOptions)
+{
```
Comment (csyonghe) [inline source/slang/slang-type-layout.cpp:1922]: This can only be std430.

File: source/slang/slang-type-layout.cpp [hunk 3]
```diff
@@ -1816,6 +1856,103 @@ LayoutRulesImpl* MetalLayoutRulesFamilyImpl::getHitAttributesParameterRules()
...
+}
+
+LayoutRulesImpl* WGSLLayoutRulesFamilyImpl::getHitAttributesParameterRules()
+{
+    return &kGLSLHitAttributesParameterLayoutRulesImpl_;
+}
+
+LayoutRulesImpl* WGSLLayoutRulesFamilyImpl::getStructuredBufferRules(CompilerOptionSet& compilerOptions)
+{
+    if (compilerOptions.shouldUseScalarLayout())
```
Comment (csyonghe) [inline source/slang/slang-type-layout.cpp:1948]: Can only be std430.

File: tests/wgsl/texture-sampler-less.slang [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+        // W-GSL: textureSampleGrad({{\(*}}t2DArray
+        && all(Tvn(T(0)) == t2DArray.SampleGrad(float3(u2, u, 0.0f), float2(ddx, ddx), float2(ddy, ddy), int2(0, 0)))
+#endif // #if TEST_WHEN_CONSTEXPR_WORKS_FOR_OFFSET
+        ;
+
+    return result;
+}
+
+void fragMain()
```
Comment (csyonghe) [inline tests/wgsl/texture-sampler-less.slang:365]: Maybe add a check for the binding location?

Comment (jkwak-work) [reply tests/wgsl/texture-sampler-less.slang:365]: That's a good idea. I will do that.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**